contrib/endpoints/src/api_manager/service_control/aggregated.cc

```diff
@@ -341,13 +439,79 @@ Status Aggregated::GetStatistics(Statistics* esp_stat) const {
   return Status::OK;
 }
 
+template <class RequestType>
+const std::string& Aggregated::GetApiReqeustUrl(const RequestType& request) {
+  if (typeid(request) == typeid(CheckRequest)) {
+    return url_.check_url();
+  } else if (typeid(request) == typeid(AllocateQuotaRequest)) {
+    return url_.quota_url();
+  } else {
+    return url_.report_url();
+  }
+}
+
+template <class RequestType>
+int Aggregated::GetHttpRequestTimeout(const RequestType& request) {
+  int timeout_ms = 0;
+
+  // Set timeout on the request if it was so configured.
+  if (typeid(request) == typeid(CheckRequest)) {
+    timeout_ms = kCheckDefaultTimeoutInMs;
+  } else if (typeid(request) == typeid(AllocateQuotaRequest)) {
+    timeout_ms = kAllocateQuotaDefaultTimeoutInMs;
+  } else {
+    timeout_ms = kReportDefaultTimeoutInMs;
+  }
+
+  if (server_config_ != nullptr &&
+      server_config_->has_service_control_config()) {
+    const auto& config = server_config_->service_control_config();
+    if (typeid(request) == typeid(CheckRequest)) {
+      if (config.check_timeout_ms() > 0) {
+        timeout_ms = config.check_timeout_ms();
+      }
+    } else if (typeid(request) == typeid(AllocateQuotaRequest)) {
+      if (config.quota_timeout_ms() > 0) {
+        timeout_ms = config.quota_timeout_ms();
+      }
+    } else {
+      if (config.report_timeout_ms() > 0) {
+        timeout_ms = config.report_timeout_ms();
+      }
+    }
+  }
+
+  return timeout_ms;
+}
+
+template <class RequestType>
+const std::string& Aggregated::GetAuthToken(const RequestType& request) {
+  if (sa_token_) {
+    if (typeid(RequestType) == typeid(AllocateQuotaRequest)) {
+      return sa_token_->GetAuthToken(
+          auth::ServiceAccountToken::JWT_TOKEN_FOR_QUOTA_CONTROL);
+    } else {
+      return sa_token_->GetAuthToken(
+          auth::ServiceAccountToken::JWT_TOKEN_FOR_SERVICE_CONTROL);
+    }
+  } else {
+    static std::string empty;
+    return empty;
+  }
+}
+
 template <class RequestType, class ResponseType>
 void Aggregated::Call(const RequestType& request, ResponseType* response,
                       TransportDoneFunc on_done,
                       cloud_trace::CloudTraceSpan* parent_span) {
   std::shared_ptr<cloud_trace::CloudTraceSpan> trace_span(
       CreateChildSpan(parent_span, "Call ServiceControl server"));
-  std::unique_ptr<HTTPRequest> http_request(new HTTPRequest([response, on_done,
+
+  const std::string& url = GetApiReqeustUrl(request);
+  TRACE(trace_span) << "Http request URL: " << url;
```

Comment:
If the Aggregated is a template class, then it would work. But those functions are template function. Without the argument, there is no way to pass the RequestType.